### PR TITLE
Feature: Extend WebSocket Configuration

### PR DESCRIPTION
**Pull request by tusharmath**

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/socket/Socket.scala [hunk 1]
```diff
@@ -20,20 +24,81 @@ object Socket {
...
+  case class OnClose[R](onClose: Connection => ZIO[R, Nothing, Unit])                    extends Socket[R, Nothing]
+  case class Concat[R, E](a: Socket[R, E], b: Socket[R, E])                              extends Socket[R, E]
+
+  sealed trait ProtocolConfig extends Socket[Any, Nothing] { self =>
+    def apply(builder: JWebSocketServerProtocolConfig.Builder): JWebSocketServerProtocolConfig.Builder =
+      ProtocolConfig.builder(self, builder)
+  }
+  object ProtocolConfig {
+    case class SubProtocol(name: String)                   extends ProtocolConfig
+    case object CheckStartsWith                            extends ProtocolConfig
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/socket/Socket.scala:42]: do we need this setting?

File: zio-http/src/main/scala/zhttp/socket/Socket.scala [hunk 2]
```diff
@@ -20,20 +24,81 @@ object Socket {
     onMessage: WebSocketFrame => ZStream[R, E, WebSocketFrame] = (_: WebSocketFrame) => ZStream.empty,
     onError: Throwable => ZIO[R, Nothing, Unit] = (_: Throwable) => ZIO.unit,
     onClose: Connection => ZIO[R, Nothing, Unit] = (_: Connection) => ZIO.unit,
-    config: JWebSocketServerProtocolConfig.Builder = JWebSocketServerProtocolConfig.newBuilder(),
+    config: JWebSocketServerProtocolConfig.Builder,
   )
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/socket/Socket.scala:28]: Logic to build settings can be moved out of settings.

File: zio-http/src/main/scala/zhttp/socket/Socket.scala [hunk 3]
```diff
@@ -20,20 +24,81 @@ object Socket {
     onMessage: WebSocketFrame => ZStream[R, E, WebSocketFrame] = (_: WebSocketFrame) => ZStream.empty,
     onError: Throwable => ZIO[R, Nothing, Unit] = (_: Throwable) => ZIO.unit,
     onClose: Connection => ZIO[R, Nothing, Unit] = (_: Connection) => ZIO.unit,
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/socket/Socket.scala:26]: We also need `onTimeout` handler.

**Review by amitksingh1490 — APPROVED** [review]

File: zio-http/src/main/scala/zhttp/socket/Socket.scala [hunk 4]
```diff
@@ -58,23 +74,69 @@ object Socket {
...
+
+  /**
+   * Close frames should be forwarded
+   */
+  def forwardCloseFrames: Socket[Any, Nothing] = ProtocolConfig.ForwardCloseFrames
+
+  /**
+   * Close frame to send, when close frame was not send manually.
+   */
+  def sendCloseFrame(status: CloseStatus): Socket[Any, Nothing] = ProtocolConfig.SendCloseFrame(status)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/socket/Socket.scala:103]: support custom status codes and reasons.

**Review by amitksingh1490 — APPROVED** [review]

**Pull request merged**